### PR TITLE
Implement basic OpenTelemetry support

### LINKED ISSUE #344: Start using the OpenTelemetry crate for telemtry
https://github.com/open-telemetry/opentelemetry-rust

Also: don't forget to update the docs/README to explain how to use it with datadog/newrelic/sentry.

More info / related attempts:
* https://github.com/tokio-rs/axum/issues/750
* https://github.com/tokio-rs/axum/pull/769
* https://github.com/tower-rs/tower-http/pull/180
* https://github.com/EmbarkStudios/server-framework/blob/main/src/middleware/trace/opentelemetry.rs
* Maybe https://github.com/davidB/sandbox_axum_observability

### PR DESCRIPTION
**Pull request by svix-daniel**
This PR integrates OpenTelemetry support and creates middleware for Axum that ensures that the spans created for each request follow OpenTelemetry conventions.

Fixes #344

### CODE REVIEW

**Review by svix-dylan — APPROVED** [review]

File: server/svix-server/src/main.rs [hunk 1]
```diff
@@ -90,19 +94,58 @@ async fn main() {
...
+            let stdout_layer = tracing_subscriber::fmt::layer()
                 .event_format(fmt)
-                .fmt_fields(json_fields)
-                .with_env_filter(filter)
+                .fmt_fields(json_fields);
+
+            tracing_subscriber::Registry::default()
+                .with(otel_layer)
+                .with(stdout_layer)
+                .with(tracing_subscriber::EnvFilter::from_default_env())
```
Comment (tasn) [inline server/svix-server/src/main.rs:148]: So it does both stdout and opentelemetry? I almost think it should be one or the other, no? I really have no idea. Maybe ask @svix-james, I think he should know better.

I'm fine with having both, I was just wondering.

File: server/svix-server/src/main.rs [hunk 2]
```diff
@@ -90,19 +94,58 @@ async fn main() {
         );
     }
 
+    let otel_layer = cfg.opentelemetry_address.as_ref().map(|addr| {
```
Comment (tasn) [inline server/svix-server/src/main.rs:97]: Shouldn't otel_layer be optional and only added if needed?

File: server/svix-server/src/main.rs [hunk 3]
```diff
@@ -90,19 +94,58 @@ async fn main() {
...
+    // either formatted normally or in JSON format
     match cfg.log_format {
         cfg::LogFormat::Default => {
-            tracing_subscriber::fmt::init();
+            let stdout_layer = tracing_subscriber::fmt::layer();
+            tracing_subscriber::Registry::default()
+                .with(otel_layer)
+                .with(stdout_layer)
+                .with(tracing_subscriber::EnvFilter::from_default_env())
+                .init();
```
Comment (tasn) [inline server/svix-server/src/main.rs:135]: This looks like exactly the same code as below, shouldn't this be shared between the two log formats?
I'm thinking:
```rust
// Is it stdout or stderr?
let stdout_layer = match cfg.log_format {
  ...
};
// Probably also dependent on whether the otel_layer exists
tracing_subscriber::Registry::default()
                .with(otel_layer)
                .with(stdout_layer)
                .with(tracing_subscriber::EnvFilter::from_default_env())
                .init();
```

**Comment by tasn** [comment]
I haven't really reviewed the code, as Dylan already has, I just reviewed the docs + main.rs.

File: server/svix-server/src/main.rs (hunk 2 above)
Comment (svix-daniel) [reply server/svix-server/src/main.rs:97]: `otel_layer` here is an `Option` as is the `opentelemetry_address`.

File: server/svix-server/src/main.rs (hunk 3 above)
Comment (svix-daniel) [reply server/svix-server/src/main.rs:135]: The `tracing_subscriber`layers are different types depending on log format, so the initialization has to occur there.

File: server/svix-server/src/main.rs (hunk 2 above)
Comment (tasn) [reply server/svix-server/src/main.rs:97]: Oh, adding layers supports options? (the `with()` below)

File: server/svix-server/src/main.rs (hunk 3 above)
Comment (tasn) [reply server/svix-server/src/main.rs:135]: It looks identical to me. The only difference I see is stdout_layer, which my comment above addresses.

File: server/svix-server/src/main.rs (hunk 2 above)
Comment (svix-daniel) [reply server/svix-server/src/main.rs:97]: Yup!

File: server/svix-server/src/lib.rs [hunk 4]
```diff
@@ -35,12 +38,14 @@ pub mod redis;
 pub mod v1;
 pub mod worker;
 
+#[tracing::instrument(name = "app_start", level = "trace")]
```
Comment (tasn) [inline server/svix-server/src/lib.rs:41]:
```suggestion
#[tracing::instrument(name = "app_start", level = "trace", skipall)]
```
Same for the rest of the functions.

Also, I'm not quite sure why we would want to instrument these functions in the first place. I get instrumenting the migration one, but why the rest of these?

File: server/svix-server/src/db/mod.rs [hunk 5]
```diff
@@ -10,6 +10,7 @@ pub mod models;
 
 static MIGRATIONS: sqlx::migrate::Migrator = sqlx::migrate!();
 
+#[tracing::instrument(level = "trace")]
```
Comment (tasn) [inline server/svix-server/src/db/mod.rs:13]: Here too, don't instrument parameters...

**Review by tasn — APPROVED** [review]

**Pull request merged**